pkg/resources/grant_privileges_to_database_role_acceptance_test.go

```diff
@@ -0,0 +1,806 @@
+			},
+			{
+				ConfigDirectory: config.TestNameDirectory(),
+				ConfigVariables: configVariables(true),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectNonEmptyPlan(),
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "always_apply", "true"),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("%s|false|true|ALL|OnDatabase|%s", databaseRoleName, databaseName)),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+			{
+				ConfigDirectory: config.TestNameDirectory(),
+				ConfigVariables: configVariables(true),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectNonEmptyPlan(),
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "always_apply", "true"),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("%s|false|true|ALL|OnDatabase|%s", databaseRoleName, databaseName)),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+			{
+				ConfigDirectory: config.TestNameDirectory(),
+				ConfigVariables: configVariables(false),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PostApplyPostRefresh: []plancheck.PlanCheck{
+						plancheck.ExpectEmptyPlan(),
+					},
+				},
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "always_apply", "false"),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("%s|false|false|ALL|OnDatabase|%s", databaseRoleName, databaseName)),
+				),
+			},
+		},
+	})
+}
+
+func createDatabaseRoleOutsideTerraform(t *testing.T, name string) {
+	t.Helper()
+	client, err := sdk.NewDefaultClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := context.Background()
+	databaseRoleId := sdk.NewDatabaseObjectIdentifier(acc.TestDatabaseName, name)
+	if err := client.DatabaseRoles.Create(ctx, sdk.NewCreateDatabaseRoleRequest(databaseRoleId).WithOrReplace(true)); err != nil {
+		t.Fatal(fmt.Errorf("error database role (%s): %w", databaseRoleId.FullyQualifiedName(), err))
+	}
+}
+
+// queriedPrivilegesEqualTo will check if all the privileges specified in the argument are granted in Snowflake.
+// Any additional grants (other than usage and ownership) will be treated as an error.
+func queriedPrivilegesEqualTo(databaseRoleName sdk.DatabaseObjectIdentifier, privileges ...string) func(s *terraform.State) error {
+	return func(s *terraform.State) error {
+		db := acc.TestAccProvider.Meta().(*sql.DB)
+		ctx := context.Background()
+		client := sdk.NewClientFromDB(db)
+		grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
+			To: &sdk.ShowGrantsTo{
+				DatabaseRole: databaseRoleName,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		for _, grant := range grants {
+			if !slices.Contains(privileges, grant.Privilege) && grant.Privilege != "USAGE" && grant.Privilege != "OWNERSHIP" {
```

Comment:
it would be nice to log here in case of usage/ownership that we do not expect the grant to be on the list. But it's just a nice-to-have, not a blocker.

Reply:
not sure what you mean by that, but I think I would leave it as is. We expect USAGE and OWNERSHIP to be always there that's why we need two additional cases. USAGE is something that the user could explicitly grant, so it could be on that list (but I wouldn't see the point of that). The other thing we could say about OWNERSHIP is that it's protected on the schema level, so I don't think anyone would specify that in the privileges ([]string in the function argument).

Reply:
I understand the nature of the additional cases. I only asked for a simple log for the devs, when we hit either of them (to better know what's happening in the test). :)

Reply:
kk, added
```go
log.Printf("Skipping check for %s privilege as its one of the privileges that are implicitly granted by Snowflake", grant.Privilege)

```